doc/man/uwildmat.3
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.32
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "uwildmat 3"
.TH uwildmat 3 "2008-04-06" "INN 2.4.6" "InterNetNews Documentation"
.SH "NAME"
uwildmat, uwildmat_simple, uwildmat_poison \- Perform wildmat matching
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fB#include <libinn.h>\fR
.PP
\&\fBbool uwildmat(const char *\fR\fItext\fR\fB, const char *\fR\fIpattern\fR\fB);\fR
.PP
\&\fBbool uwildmat_simple(const char *\fR\fItext\fR\fB, const char *\fR\fIpattern\fR\fB);\fR
.PP
\&\fBenum uwildmat uwildmat_poison(const char *\fR\fItext\fR\fB,
const char *\fR\fIpattern\fR\fB);\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBuwildmat\fR compares \fItext\fR against the wildmat expression \fIpattern\fR,
returning true if and only if the expression matches the text.  \f(CW\*(C`@\*(C'\fR has
no special meaning in \fIpattern\fR when passed to \fBuwildmat\fR.  Both \fItext\fR
and \fIpattern\fR are assumed to be in the \s-1UTF\-8\s0 character encoding, although
malformed \s-1UTF\-8\s0 sequences are treated in a way that attempts to be mostly
compatible with single-octet character sets like \s-1ISO\s0 8859\-1.  (In other
words, if you try to match \s-1ISO\s0 8859\-1 text with these routines everything
should work as expected unless the \s-1ISO\s0 8859\-1 text contains valid \s-1UTF\-8\s0
sequences, which thankfully is somewhat rare.)
.PP
\&\fBuwildmat_simple\fR is identical to \fBuwildmat\fR except that neither \f(CW\*(C`!\*(C'\fR
nor \f(CW\*(C`,\*(C'\fR have any special meaning and \fIpattern\fR is always treated as a
single pattern.  This function exists solely to support legacy interfaces
like \s-1NNTP\s0's \s-1XPAT\s0 command, and should be avoided when implementing new
features.
.PP
\&\fBuwildmat_poison\fR works similarly to \fBuwildmat\fR, except that \f(CW\*(C`@\*(C'\fR as the
first character of one of the patterns in the expression (see below)
\&\*(L"poisons\*(R" the match if it matches.  \fBuwildmat_poison\fR returns
\&\fB\s-1UWILDMAT_MATCH\s0\fR if the expression matches the text, \fB\s-1UWILDMAT_FAIL\s0\fR if
it doesn't, and \fB\s-1UWILDMAT_POISON\s0\fR if the expression doesn't match because
a poisoned pattern matched the text.  These enumeration constants are
defined in the \fBlibinn.h\fR header.
.SH "WILDMAT EXPRESSIONS"
.IX Header "WILDMAT EXPRESSIONS"
A wildmat expression follows rules similar to those of shell filename
wildcards but with some additions and changes.  A wildmat \fIexpression\fR is
composed of one or more wildmat \fIpatterns\fR separated by commas.  Each
character in the wildmat pattern matches a literal occurance of that same
character in the text, with the exception of the following metacharacters:
.IP "?" 8
Matches any single character (including a single \s-1UTF\-8\s0 multibyte
character, so \f(CW\*(C`?\*(C'\fR can match more than one byte).
.IP "*\&" 8
Matches any sequence of zero or more characters.
.IP "\e" 8
.IX Item ""
Turns off any special meaning of the following character; the following
character will match itself in the text.  \f(CW\*(C`\e\*(C'\fR will escape any character,
including another backslash or a comma that otherwise would separate a
pattern from the next pattern in an expression.  Note that \f(CW\*(C`\e\*(C'\fR is not
special inside a character range (no metacharacters are).
.IP "[...]" 8
A character set, which matches any single character that falls within that
set.  The presence of a character between the brackets adds that character
to the set; for example, \f(CW\*(C`[amv]\*(C'\fR specifies the set containing the
characters \f(CW\*(C`a\*(C'\fR, \f(CW\*(C`m\*(C'\fR, and \f(CW\*(C`v\*(C'\fR.  A range of characters may be specified
using \f(CW\*(C`\-\*(C'\fR; for example, \f(CW\*(C`[0\-5abc]\*(C'\fR is equivalent to \f(CW\*(C`[012345abc]\*(C'\fR.  The
order of characters is as defined in the \s-1UTF\-8\s0 character set, and if the
start character of such a range falls after the ending character of the
range in that ranking the results of attempting a match with that pattern
are undefined.
.Sp
In order to include a literal \f(CW\*(C`]\*(C'\fR character in the set, it must be the
first character of the set (possibly following \f(CW\*(C`^\*(C'\fR); for example, \f(CW\*(C`[]a]\*(C'\fR
matches either \f(CW\*(C`]\*(C'\fR or \f(CW\*(C`a\*(C'\fR.  To include a literal \f(CW\*(C`\-\*(C'\fR character in the
set, it must be either the first or the last character of the set.
Backslashes have no special meaning inside a character set, nor do any
other of the wildmat metacharacters.
.IP "[^...]" 8
A negated character set.  Follows the same rules as a character set above,
but matches any character \fBnot\fR contained in the set.  So, for example,
\&\f(CW\*(C`[^]\-]\*(C'\fR matches any character except \f(CW\*(C`]\*(C'\fR and \f(CW\*(C`\-\*(C'\fR.
.PP
In addition, \f(CW\*(C`!\*(C'\fR (and possibly \f(CW\*(C`@\*(C'\fR) have special meaning as the first
character of a pattern; see below.
.PP
When matching a wildmat expression against some text, each comma-separated
pattern is matched in order from left to right.  In order to match, the
pattern must match the whole text; in regular expression terminology, it's
implicitly anchored at both the beginning and the end.  For example, the
pattern \f(CW\*(C`a\*(C'\fR matches only the text \f(CW\*(C`a\*(C'\fR; it doesn't match \f(CW\*(C`ab\*(C'\fR or \f(CW\*(C`ba\*(C'\fR
or even \f(CW\*(C`aa\*(C'\fR.  If none of the patterns match, the whole expression
doesn't match.  Otherwise, whether the expression matches is determined
entirely by the rightmost matching pattern; the expression matches the
text if and only if the rightmost matching pattern is not negated.
.PP
For example, consider the text \f(CW\*(C`news.misc\*(C'\fR.  The expression \f(CW\*(C`*\*(C'\fR matches
this text, of course, as does \f(CW\*(C`comp.*,news.*\*(C'\fR (because the second pattern
matches).  \f(CW\*(C`news.*,!news.misc\*(C'\fR does not match this text because both
patterns match, meaning that the rightmost takes precedence, and the
rightmost matching pattern is negated.  \f(CW\*(C`news.*,!news.misc,*.misc\*(C'\fR does
match this text, since the rightmost matching pattern is not negated.
.PP
Note that the expression \f(CW\*(C`!news.misc\*(C'\fR can't match anything.  Either the
pattern doesn't match, in which case no patterns match and the expression
doesn't match, or the pattern does match, in which case because it's
negated the expression doesn't match.  \f(CW\*(C`*,!news.misc\*(C'\fR, on the other hand,
is a useful pattern that matches anything except \f(CW\*(C`news.misc\*(C'\fR.
.PP
\&\f(CW\*(C`!\*(C'\fR has significance only as the first character of a pattern; anywhere
else in the pattern, it matches a literal \f(CW\*(C`!\*(C'\fR in the text like any other
non\-metacharacter.
.PP
If the \fBuwildmat_poison\fR interface is used, then \f(CW\*(C`@\*(C'\fR behaves the same as
\&\f(CW\*(C`!\*(C'\fR except that if an expression fails to match because the rightmost
matching pattern began with \f(CW\*(C`@\*(C'\fR, \fB\s-1UWILDMAT_POISON\s0\fR is returned instead of
\&\fB\s-1UWILDMAT_FAIL\s0\fR.
.PP
If the \fBuwildmat_simple\fR interface is used, the matching rules are the
same as above except that none of \f(CW\*(C`!\*(C'\fR, \f(CW\*(C`@\*(C'\fR, or \f(CW\*(C`,\*(C'\fR have any special
meaning at all and only match those literal characters.
.SH "BUGS"
.IX Header "BUGS"
All of these functions internally convert the passed arguments to const
unsigned char pointers.  The only reason why they take regular char
pointers instead of unsigned char is for the convenience of \s-1INN\s0 and other
callers that may not be using unsigned char everywhere they should.  In a
future revision, the public interface should be changed to just take
unsigned char pointers.
.SH "HISTORY"
.IX Header "HISTORY"
Written by Rich \f(CW$alz\fR <rsalz@uunet.uu.net> in 1986, and posted to Usenet
several times since then, most notably in comp.sources.misc in
March, 1991.
.PP
Lars Mathiesen <thorinn@diku.dk> enhanced the multi-asterisk failure
mode in early 1991.
.PP
Rich and Lars increased the efficiency of star patterns and reposted it to
comp.sources.misc in April, 1991.
.PP
Robert Elz <kre@munnari.oz.au> added minus sign and close bracket handling
in June, 1991.
.PP
Russ Allbery <rra@stanford.edu> added support for comma-separated patterns
and the \f(CW\*(C`!\*(C'\fR and \f(CW\*(C`@\*(C'\fR metacharacters to the core wildmat routines in July,
2000.  He also added support for \s-1UTF\-8\s0 characters, changed the default
behavior to assume that both the text and the pattern are in \s-1UTF\-8\s0, and
largely rewrote this documentation to expand and clarify the description
of how a wildmat expression matches.
.PP
Please note that the interfaces to these functions are named \fBuwildmat\fR
and the like rather than \fBwildmat\fR to distinguish them from the
\&\fBwildmat\fR function provided by Rich \f(CW$alz\fR's original implementation.
While this code is heavily based on Rich's original code, it has
substantial differences, including the extension to support \s-1UTF\-8\s0
characters, and has noticable functionality changes.  Any bugs present in
it aren't Rich's fault.
.PP
$Id$
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIgrep\fR\|(1), \fIfnmatch\fR\|(3), \fIregex\fR\|(3), \fIregexp\fR\|(3).


NEWS
Changes in 2.4.6

    * Fixed the segfault of the radius authenticator when none of the radius
      servers respond.  Thanks to Matija Nalis for this patch.

    * Fixed a lost initialization in buffindexed, which resolves a potential
      segfault, thanks to a patch by Kirill Berezin.

    * INN now properly supports Perl 5.10.0; Perl filters were causing innd
      to segfault on a few systems like FreeBSD.

    * Fixed a long-standing bug which affected Perl hooks for innd: the
      variable containing the body of an article was not properly created,
      which caused regular expressions matching new lines to fail.  It
      especially affected filters like Cleanfeed which sometimes failed to
      detect unwanted articles.

      To fix that issue, Julien Elie added the use of a shared string,
      available since Perl 5.7.2, with a fall back to a slower but reliable
      copy of such bodies in case the function is not available.  Using a
      Perl version superior to 5.7.2 is therefore recommended.

    * Fixed two bugs which could prevent nnrpd from being run as a daemon in
      FreeBSD.  Thanks to Johan van Selst for having identified the problem
      and to Kai Gallasch for having provided a testing FreeBSD server.  The
      listening address was not initialized to "::0" or 0.0.0.0 when the -b
      flag was not used and an incorrect size was given when IPv6 was
      enabled and the binding done using IPv4.

    * Some annoying assertion failures occurring in innfeed have been fixed
      by Russ Allbery and Julien Elie.

    * Fixed a bug in mod-active for aliased newsgroups.  Only "=" was
      written to the active file.  Thanks to D. Stussy for this patch.

    * Fixed a bug which caused innd not to honour the Ad flag in newsfeeds.

    * Fixed a bug in the IP address displayed for "localhost" in innd's
      status file.  It was not correctly initialized.

    * Fixed a permission issue:  XHDR and XPAT were not checking the rights
      the user had to read articles when accessing them by their message-ID.

    * Fixed a bug in the replies of XHDR, XOVER and XPAT when the newsgroup
      is empty.  Two initial replies were sent instead of one:  the right
      420 code followed by a wrong 224 code.

    * When no newsgroup is selected, LISTGROUP now returns the right 412
      code (instead of 481).

    * inncheck now uses a range of permissions to see whether the file modes
      are correctly set.  Therefore, different configurations depending on
      the security the user wants to enforce on his sytem are possible.

    * A new improved version of docheckgroups is shipped with INN.  The -u
      flag permits to automatically update the newsgroups file (with a
      proper number of tabulations and an alphabetical sort), removing
      obsolete descriptions and adding new ones.  A second argument on
      command-line permits to specify which newsgroups should not be
      checked, so as not to treat them.

    * An *email=* keyword has been added by James Ralston to news.daily in
      order to supply another mail address than the one set at configure
      time for Usenet daily reports.

    * An updated moderators file with information about the aioe.*, perl.*
      and si.* hierarchies is provided; control.ctl is also up to date.

    * INN supports Berkeley DB 4.7, which is the recommended version to use
      owing to various bugs affecting previous versions of Berkeley DB.

    * Other minor bugs have also been fixed.

Changes in 2.4.5

    * Fixed the "alarm signal" around "SSL_read" in nnrpd:  it allows a
      proper disconnection of news clients which were previously hanging
      when posting an article through a SSL connection.  Moreover, the
      *clienttimeout* parameter now works on SSL connections.  Thanks to
      Matija Nalis for the patch.

    * SO_KEEPALIVE is now implemented for SSL TCP connections on systems
      which support it, allowing system detection and closing the dead TCP
      SSL connections automatically after system-specified time.  Thanks to
      Matija Nalis for the patch.

    * Fixed a segmentation fault when an article of a size greater than
      remaining stack is retrieved via SSL.  Thanks to Chris Caputo for this
      patch.

    * Fixed a few segfaults and bugs which affected both Python innd and
      nnrpd hooks.  They no longer check the existence of methods not used
      by the hooked script.  An issue with Python exception handling was
      also fixed, as well as a segfault fixed by Russ Allbery which happened
      whenever one closes and then reopens Python in the same process. 
      Julien Elie also fixed a bug when reloading Python filters (they were
      not always correctly reloaded) and a segfault when generating access
      groups with embedded Python filters for nnrpd.  Many thanks to David
      Hlacik for its bug reports.

    * The nnrpd.py stub file in order to test Python nnrpd hooks, as
      mentioned in their documentation, is now installed; only INN.py was
      previously installed in *pathfilter*.  Also fixed a bug in INN.py and
      add missing methods to it.

    * Fixed a long-standing bug in innreport which prevented it from
      correctly reporting nnrpd and innfeed log messages.

    * Fixed a hang in Perl hooks on (at least) HP/PA since Perl 5.10.

    * Fixed a compilation problem on some platforms because of AF_INET6
      which was not inside a HAVE_INET6 block in innfeed.

    * Fixed a bug in innfeed which contained thrice the same IPs for each
      peer; it unnecessarily slowed the peer IP rotation for innfeed. 
      Thanks, D. Stussy, for having seen that.  Miquel van Smoorenburg
      provided the patch.

    * A new *heavily* improved version of pullnews is shipped with this INN
      release.  This new version is provided by Geraint Edwards.  He added
      no more than 16 flags, fixed some bugs and integrated the backupfeed

lib/uwildmat.c
/*  $Id$
**
**  wildmat pattern matching with Unicode UTF-8 extensions.
**
**  Do shell-style pattern matching for ?, \, [], and * characters.  Might not
**  be robust in face of malformed patterns; e.g., "foo[a-" could cause a
**  segmentation violation.  It is 8-bit clean.  (Robustness hopefully fixed
**  July 2000; all malformed patterns should now just fail to match anything.)
**
**  Original by Rich $alz, mirror!rs, Wed Nov 26 19:03:17 EST 1986.
**  Rich $alz is now <rsalz@osf.org>.
**
**  April, 1991:  Replaced mutually-recursive calls with in-line code for the
**  star character.
**
**  Special thanks to Lars Mathiesen <thorinn@diku.dk> for the ABORT code.
**  This can greatly speed up failing wildcard patterns.  For example:
**
**	pattern: -*-*-*-*-*-*-12-*-*-*-m-*-*-*
**	text 1:	 -adobe-courier-bold-o-normal--12-120-75-75-m-70-iso8859-1
**	text 2:	 -adobe-courier-bold-o-normal--12-120-75-75-X-70-iso8859-1
**
**  Text 1 matches with 51 calls, while text 2 fails with 54 calls.  Without
**  the ABORT code, it takes 22310 calls to fail.  Ugh.  The following
**  explanation is from Lars:
**
**  The precondition that must be fulfilled is that DoMatch will consume at
**  least one character in text.  This is true if *p is neither '*' nor '\0'.)
**  The last return has ABORT instead of false to avoid quadratic behaviour in
**  cases like pattern "*a*b*c*d" with text "abcxxxxx".  With false, each
**  star-loop has to run to the end of the text; with ABORT only the last one
**  does.
**
**  Once the control of one instance of DoMatch enters the star-loop, that
**  instance will return either true or ABORT, and any calling instance will
**  therefore return immediately after (without calling recursively again).
**  In effect, only one star-loop is ever active.  It would be possible to
**  modify the code to maintain this context explicitly, eliminating all
**  recursive calls at the cost of some complication and loss of clarity (and
**  the ABORT stuff seems to be unclear enough by itself).  I think it would
**  be unwise to try to get this into a released version unless you have a
**  good test data base to try it out on.
**
**  June, 1991:  Robert Elz <kre@munnari.oz.au> added minus and close bracket
**  handling for character sets.
**
**  July, 2000:  Largely rewritten by Russ Allbery <rra@stanford.edu> to add
**  support for ',', '!', and optionally '@' to the core wildmat routine.
**  Broke the character class matching into a separate function for clarity
**  since it's infrequently used in practice, and added some simple lookahead
**  to significantly decrease the recursive calls in the '*' matching code.
**  Added support for UTF-8 as the default character set for any high-bit
**  characters.
**
**  For more information on UTF-8, see RFC 2279.
**
**  Please note that this file is intentionally written so that conditionally
**  executed expressions are on separate lines from the condition to
**  facilitate analysis of the coverage of the test suite using purecov.
**  Please preserve this.  As of March 11, 2001, purecov reports that the
**  accompanying test suite achieves 100% coverage of this file.
*/

#include "config.h"
#include "clibrary.h"
#include "inn/libinn.h"

#define ABORT -1

/* Whether or not an octet looks like the start of a UTF-8 character. */
#define ISUTF8(c)       (((c) & 0xc0) == 0xc0)


/*
**  Determine the length of a non-ASCII character in octets (for advancing
**  pointers when skipping over characters).  Takes a pointer to the start of
**  the character and to the last octet of the string.  If end is NULL, expect
**  the string pointed to by start to be nul-terminated.  If the character is
**  malformed UTF-8, return 1 to treat it like an eight-bit local character.
*/
static int
utf8_length(const unsigned char *start, const unsigned char *end)
{
    unsigned char mask = 0x80;
    const unsigned char *p;
    int length = 0;
    int left;

    for (; mask > 0 && (*start & mask) == mask; mask >>= 1)
        length++;
    if (length < 2 || length > 6)
        return 1;
    if (end != NULL && (end - start + 1) < length)
        return 1;
    left = length - 1;
    p = start + 1;
    for (p = start + 1; left > 0 && (*p & 0xc0) == 0x80; p++)
        left--;
    return (left == 0) ? length : 1;
}


/*
**  Convert a UTF-8 character to UCS-4.  Takes a pointer to the start of the
**  character and to the last octet of the string, and to a uint32_t into
**  which to put the decoded UCS-4 value.  If end is NULL, expect the string
**  pointed to by start to be nul-terminated.  Returns the number of octets in
**  the UTF-8 encoding.  If the UTF-8 character is malformed, set result to
**  the decimal value of the first octet; this is wrong, but it will generally
**  cause the rest of the wildmat matching to do the right thing for non-UTF-8
**  input.
*/
static int
utf8_decode(const unsigned char *start, const unsigned char *end,
            uint32_t *result)
{
    uint32_t value = 0;
    int length, i;
    const unsigned char *p = start;
    unsigned char mask;

    length = utf8_length(start, end);
    if (length < 2) {
        *result = *start;
        return 1;
    }
    mask = (1 << (7 - length)) - 1;
    value = *p & mask;
    p++;
    for (i = length - 1; i > 0; i--) {
        value = (value << 6) | (*p & 0x3f);
        p++;
    }
    *result = value;
    return length;
}


/*
**  Match a character class against text, a UCS-4 character.  start is a
**  pointer to the first character of the character class, end a pointer to
**  the last.  Returns whether the class matches that character.
*/
static bool
match_class(uint32_t text, const unsigned char *start,
            const unsigned char *end)
{
    bool reversed, allowrange;
    const unsigned char *p = start;
    uint32_t first, last;

    /* Check for an inverted character class (starting with ^).  If the
       character matches the character class, we return !reversed; that way,
       we return true if it's a regular character class and false if it's a
       reversed one.  If the character doesn't match, we return reversed. */
    reversed = (*p == '^');
    if (reversed)
        p++;

    /* Walk through the character class until we reach the end or find a
       match, handling character ranges as we go.  Only permit a range to
       start when allowrange is true; this allows - to be treated like a
       normal character as the first character of the class and catches
       malformed ranges like a-e-n.  We treat the character at the beginning
       of a range as both a regular member of the class and the beginning of
       the range; this is harmless (although it means that malformed ranges
       like m-a will match m and nothing else). */
    allowrange = false;
    while (p <= end) {
        if (allowrange && *p == '-' && p < end) {
            p++;
            p += utf8_decode(p, end, &last);
            if (text >= first && text <= last)
                return !reversed;
            allowrange = false;
        } else {
            p += utf8_decode(p, end, &first);
            if (text == first)
                return !reversed;
            allowrange = true;
        }
    }
    return reversed;
}


/*
**  Match the text against the pattern between start and end.  This is a
**  single pattern; a leading ! or @ must already be taken care of, and
**  commas must be dealt with outside of this routine.
*/
static int
match_pattern(const unsigned char *text, const unsigned char *start,
              const unsigned char *end)
{
    const unsigned char *q, *endclass;
    const unsigned char *p = start;
    bool ismeta;
    int matched, width;
    uint32_t c;

    for (; p <= end; p++) {
        if (!*text && *p != '*')
            return ABORT;

        switch (*p) {
        case '\\':
            if (!*++p)
                return ABORT;
            /* Fall through. */

        default:
            if (*text++ != *p)
                return false;
            break;

        case '?':
            text += ISUTF8(*text) ? utf8_length(text, NULL) : 1;
            break;

        case '*':
            /* Consecutive stars are equivalent to one.  Advance pattern to
               the character after the star. */
            for (++p; *p == '*'; p++)
                ;

            /* A trailing star will match anything. */
            if (p > end)
                return true;

            /* Basic algorithm: Recurse at each point where the * could
               possibly match.  If the match succeeds or aborts, return
               immediately; otherwise, try the next position.

               Optimization: If the character after the * in the pattern
               isn't a metacharacter (the common case), then the * has to
               consume characters at least up to the next occurance of that
               character in the text.  Scan forward for those points rather
               than recursing at every possible point to save the extra
               function call overhead. */
            ismeta = (*p == '[' || *p == '?' || *p == '\\');
            while (*text) {
                width = ISUTF8(*text) ? utf8_length(text, NULL) : 1;
                if (ismeta) {
                    matched = match_pattern(text, p, end);
                    text += width;
                } else {
                    while (*text && *text != *p) {
                        text += width;
                        width = ISUTF8(*text) ? utf8_length(text, NULL) : 1;
                    }
                    if (!*text)
                        return ABORT;
                    matched = match_pattern(++text, p + 1, end);
                }
                if (matched != false)
                    return matched;
            }
            return ABORT;

        case '[':
            /* Find the end of the character class, making sure not to pick
               up a close bracket at the beginning of the class. */
            p++;
            q = p + (*p == '^') + 1;
            if (q > end)
                return ABORT;
            endclass = memchr(q, ']', (size_t) (end - q + 1));
            if (!endclass)
                return ABORT;

            /* Do the heavy lifting in another function for clarity, since
               character classes are an uncommon case. */
            text += utf8_decode(text, NULL, &c);
            if (!match_class(c, p, endclass - 1))
                return false;
            p = endclass;
            break;
        }
    }

    return (*text == '\0');
}


/*
**  Takes text and a wildmat expression; a wildmat expression is a
**  comma-separated list of wildmat patterns, optionally preceeded by ! to
**  invert the sense of the expression.  Returns WILDMAT_MATCH if that
**  expression matches the text, WILDMAT_FAIL otherwise.  If allowpoison is
**  set, allow @ to introduce a poison expression (the same as !, but if it
**  triggers the failed match the routine returns WILDMAT_POISON instead).
*/
static enum uwildmat
match_expression(const unsigned char *text, const unsigned char *start,
                 bool allowpoison)
{
    const unsigned char *end, *split;
    const unsigned char *p = start;
    bool reverse, escaped;
    bool match = false;
    bool poison = false;
    bool poisoned = false;

    /* Handle the empty expression separately, since otherwise end will be
       set to an invalid pointer. */
    if (!*p)
        return !*text ? UWILDMAT_MATCH : UWILDMAT_FAIL;
    end = start + strlen((const char *) start) - 1;

    /* Main match loop.  Find each comma that separates patterns, and attempt 
       to match the text with each pattern in order.  The last matching
       pattern determines whether the whole expression matches. */
    for (; p <= end + 1; p = split + 1) {
        if (allowpoison)
            poison = (*p == '@');
        reverse = (*p == '!') || poison;
        if (reverse)
            p++;

        /* Find the first unescaped comma, if any.  If there is none, split
           will be one greater than end and point at the nul at the end of
           the string. */
        for (escaped = false, split = p; split <= end; split++) {
            if (*split == '[') {
                split++;
                if (*split == ']')
                    split++;
                while (split <= end && *split != ']')
                    split++;
            }
            if (*split == ',' && !escaped)
                break;
            escaped = (*split == '\\') ? !escaped : false;
        }

        /* Optimization: If match == !reverse and poison == poisoned, this
           pattern can't change the result, so don't do any work. */
        if (match == !reverse && poison == poisoned)
            continue;
        if (match_pattern(text, p, split - 1) == true) {
            poisoned = poison;
            match = !reverse;
        }
    }
    if (poisoned)
        return UWILDMAT_POISON;
    return match ? UWILDMAT_MATCH : UWILDMAT_FAIL;
}


/*
**  User-level routine used for wildmats where @ should be treated as a
**  regular character.
*/
bool
uwildmat(const char *text, const char *pat)
{
    const unsigned char *utext = (const unsigned char *) text;
    const unsigned char *upat = (const unsigned char *) pat;

    if (upat[0] == '*' && upat[1] == '\0')
        return true;
    else
        return (match_expression(utext, upat, false) == UWILDMAT_MATCH);
}


/*
**  User-level routine used for wildmats that support poison matches.
*/
enum uwildmat
uwildmat_poison(const char *text, const char *pat)
{
    const unsigned char *utext = (const unsigned char *) text;
    const unsigned char *upat = (const unsigned char *) pat;

    if (upat[0] == '*' && upat[1] == '\0')
        return UWILDMAT_MATCH;
    else
        return match_expression(utext, upat, true);
}


/*
**  User-level routine for simple expressions (neither , nor ! are special).
*/
bool
uwildmat_simple(const char *text, const char *pat)
{
    const unsigned char *utext = (const unsigned char *) text;
    const unsigned char *upat = (const unsigned char *) pat;
    size_t length;

    if (upat[0] == '*' && upat[1] == '\0')
        return true;
    else {
        length = strlen(pat);
        return (match_pattern(utext, upat, upat + length - 1) == true);
    }
}
